lua/entities/gmod_tardis/modules/sh_tracking.lua

```diff
@@ -0,0 +1,497 @@
+-- Tracking
+
+local TRACKING_MAX_DISTANCE_SET = 1000
+local TRACKING_MAX_DISTANCE_TARGET_MAX = 1500
+local TRACKING_MAX_DISTANCE_NO_LOS = 2000
+local TRACKING_MAX_DISTANCE_TRACE = 10000
+
+local TRACKING_DEBUG = false
+
+TARDIS:AddKeyBind("tracking",{
+    name="Tracking",
+    section="Flight",
+    func=function(self,down,ply)
+        local pilot = self:GetData("pilot")
+        if down and self:CallHook("CanTrack") == false then return end
+        if ply==pilot then
+            if down then
+                self:SetData("tracking-trace",true)
+            else
+                self:SendMessage("tracking-set", {self:GetData("tracking-ent")})
+            end
+        end
+        if not down then
+            self:SetData("tracking-trace",false)
+            self:SetData("tracking-ent",nil)
+        end
+    end,
+    key=KEY_X,
+    clientonly=true,
+    exterior=true
+})
+
+TARDIS:AddKeyBind("tracking-rotation",{
+    name="TrackRotation",
+    section="Flight",
+    func=function(self,down,ply)
+        local pilot = self:GetData("pilot")
+        if not self:GetTracking() or ply ~= pilot or not down then return end
+        self:ToggleTrackRotation()
+        TARDIS:StatusMessage(ply, "Controls.Tracking.Rotation", self:GetTrackRotation())
+    end,
+    key=KEY_T,
+    serveronly=true,
+    exterior=true
+})
+
+if SERVER then
+    function ENT:GetTracking()
+        return self:GetData("tracking-ent")
+    end
+
+    function ENT:SetTracking(ent,ply)
+        local wasTrackingEnt = self:GetData("tracking-ent")
+        local wasTracking = IsValid(wasTrackingEnt)
+        local isTracking = IsValid(ent)
+        local valid = true
+        if isTracking then 
+            if ent.TardisPart or ent.TardisInterior or (ent:IsPlayer() and IsValid(TARDIS:GetInteriorEnt(ent))) then
+                if IsValid(ply) then
+                    TARDIS:ErrorMessage(ply, "Controls.Tracking.InteriorFail")
+                end
```

Comment:
we don't need to check IsValid(ply) since that check exists at sh_notifications.lua:15

Reply:
https://github.com/MattJeanes/TARDIS/pull/969/commits/5b22f4263b0ffd22a65c51f7013232fc202a9594